libsel4vmmplatsupport/src/drivers/virtio_block_emul.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2019, DornerWorks
+ *
+ * This software may be distributed and modified according to the terms of
+ * the GNU General Public License version 2. Note that NO WARRANTY is provided.
+ * See "LICENSE_GPLv2.txt" for details.
+ *
+ * @TAG(DORNERWORKS_GPL)
+ */
+
+#include <sel4vmmplatsupport/drivers/virtio_pci_emul.h>
+#include <stdbool.h>
+
+#include "virtio_emul_helpers.h"
+
+#define BUF_SIZE 8192
+#define MAX_DATA_BUF_SIZE 4096
+
+typedef struct blkif_virtio_emul_internal {
+    struct disk_driver driver;
+    struct virtio_blk_config cfg;
+    ps_dma_man_t dma_man;
+} blkif_virtio_emul_internal_t;
+
+typedef struct emul_tx_cookie {
+    uint16_t desc_head;
+    void *vaddr;
+} emul_tx_cookie_t;
+
+static void complete_virtio_blk_request(void *iface, void *cookie)
+{
+    virtio_emul_t *emul = (virtio_emul_t *) iface;
+    blkif_virtio_emul_internal_t *blk = emul->internal;
+    emul_tx_cookie_t *tx_cookie = (emul_tx_cookie_t *)cookie;
+    /* free the dma memory */
+    ps_dma_unpin(&blk->dma_man, tx_cookie->vaddr, BUF_SIZE);
+    ps_dma_free(&blk->dma_man, tx_cookie->vaddr, BUF_SIZE);
+    /* put the descriptor chain into the used list */
+    struct vring_used_elem used_elem = {tx_cookie->desc_head, 0};
+    ring_used_add(emul, &emul->virtq.vring[emul->virtq.queue], used_elem);
+    free(tx_cookie);
+    /* notify the guest that we have completed some of its buffers */
+    blk->driver.i_fn.raw_handleIRQ(&blk->driver, 0);
+}
+
+static void handle_virtio_blk_request(virtio_emul_t *emul)
+{
+    /* Create Local Copies of the Passed in Structure */
+    blkif_virtio_emul_internal_t *blk = (blkif_virtio_emul_internal_t *) emul->internal;
+    struct vring *vring = &emul->virtq.vring[emul->virtq.queue];
+
+    /* read the index */
+    uint16_t guest_idx = ring_avail_idx(emul, vring);
+
+    /* process what we can of the ring */
+    uint16_t idx = emul->virtq.last_idx[emul->virtq.queue];
+    uint32_t buf_len = 0;
+
+    uint64_t desc_addrs[3];
```

Comment:
Can we put this in a define at the top or in the header

Reply:
Sounds good, will put at the top